### PR TITLE
Handle client cache failure on "no space left on device"

### PR DESCRIPTION
**Pull request by apc999**
This happens when (1) cache is configured with wrong and smaller capacity, (2) for some reason disk capacity allocated to Alluxio was "stolen" by other applications, or (3) alluxio cache is under-counting the real space used on disk (e.g., due to bugs, inode overhead and etc).

This PR aim to
1. add a metrics to monitor and report when `no space left on device` happens
2. force evicting items in a retry in this case
3.  add a new flag `alluxio.user.client.cache.store.overhead` to allow users to set the space overhead when writing data to disk

Fix https://github.com/prestodb/presto/issues/15972

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java [hunk 1]
```diff
@@ -65,21 +66,24 @@ public LocalPageStore(LocalPageStoreOptions options) {
...
+        Files.createFile(p);
+      }
       // extra try to ensure output stream is closed
       try (FileOutputStream fos = new FileOutputStream(p.toFile(), false)) {
         fos.write(page);
       }
     } catch (Exception e) {
       Files.deleteIfExists(p);
+      if (e.getMessage().contains("No space left on device")) {
+        throw new ResourceExhaustedException("No space left on device");
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java:85]: Make the string a constant? Also is it guaranteed or is there a better way of referencing it?

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java [hunk 2]
```diff
@@ -348,9 +353,14 @@ private PutResult putAttempt(PageId pageId, byte[] page, CacheScope scope, Cache
...
           return PutResult.OK;
+        } catch (ResourceExhaustedException e) {
+          undoAddPage(pageId);
+          LOG.error("Failed to add page {} to pageStore: ", pageId, e);
+          Metrics.PUT_STORE_WRITE_NO_SPACE_ERRORS.inc();
+          return PutResult.NO_SPACE;
         } catch (IOException e) {
           undoAddPage(pageId);
-          LOG.error("Failed to add page {}: {}", pageId, e);
+          LOG.error("Failed to add page {} to pageStore: ", pageId, e);
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:363]: Is the notation to have `: ` before the exception, I don't think we do this in other places?

**Review by calvinjia — CHANGES_REQUESTED** [review]

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java (hunk 1 above)
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java:85]: Would it be helpful to also state the device (`rootDir`) that is full?

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java (hunk 2 above)
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:363]: I was using this convention in this class

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java (hunk 1 above)
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java:85]: done.
it is not guaranteed, this is best effort for monitoring and recover from this error

Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java:85]: done

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java [hunk 3]
```diff
@@ -308,13 +318,14 @@ private boolean putInternal(PageId pageId, byte[] page, CacheScope scope, CacheQ
...
+    } else if (result == PutResult.INSUFFICIENT_SPACE_EVICTED) {
       Metrics.PUT_INSUFFICIENT_SPACE_ERRORS.inc();
     }
     return false;
   }
 
-  private PutResult putAttempt(PageId pageId, byte[] page, CacheScope scope, CacheQuota quota) {
+  private PutResult putAttempt(PageId pageId, byte[] page, CacheScope scope, CacheQuota quota,
+      boolean forcedToEvict) {
     LOG.debug("putInternal({},{} bytes) enters", pageId, page.length);
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:329]: Should we also log `forcedToEvict` or have some metrics around that?

**Review by calvinjia — CHANGES_REQUESTED** [review]
Had a minor comment

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java (hunk 3 above)
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java:329]: @calvinjia  tracked by metrics `PUT_STORE_WRITE_NO_SPACE_ERROR`

**Comment by apc999** [comment]
@calvinjia  PTAL

**Review by calvinjia — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**